lib/wibox/init.lua

```diff
@@ -61,6 +63,106 @@ function wibox:find_widgets(x, y)
     return self._drawable:find_widgets(x, y)
 end
 
+-- Smarted version of the function found in surface.lua.
+-- As wiboxes are more flexible than clients, this allows proper clipping.
+local function apply_shape(self)
+    local shape = self._shape
+
+    if not shape then return end
+
+    local geo = self:geometry()
+    local bw = self.border_width
+
+    local img = cairo.ImageSurface(
+        cairo.Format.A1,
+        geo.width  + 2*bw,
+        geo.height + 2*bw
+    )
+    local cr = cairo.Context(img)
+
+    cr:set_operator(cairo.Operator.CLEAR)
+    cr:set_source_rgba(0,0,0,1)
+    cr:paint()
+
+    cr:set_operator(cairo.Operator.SOURCE)
+    cr:set_source_rgba(1,1,1,1)
+
+    shape(cr, geo.width+2*bw, geo.height+2*bw)
+
+    cr:fill()
+
+    self.shape_bounding = img._native
+
+    -- Now, set an empty clip. The wibox clip is drawn using cairo in
+    -- drawable.lua so anti-alising work.
+    img = cairo.ImageSurface(
+        cairo.Format.A1,
+        geo.width,
+        geo.height
+    )
+    cr = cairo.Context(img)
+
+
+    -- When there is a compositing manager, clip all the available area and
+    -- let the drawable.lua implementation do its job. However, this will leave
+    -- a 1px "black" border if there isn't any compositing.
+    -- Note that this code doesn't handle compositing manager being turned on
+    -- and off.
+    if not capi.awesome.composite_manager_running then
+        cr:translate(-bw, -bw)
+        shape(cr, geo.width+2*bw, geo.height+2*bw)
+
+        -- As shapes don't all grow radially, use this little trick to get the
+        -- 1 pixel outline of the shape (as applied above)
+        cr:set_source_rgba(1,1,1,1)
+        cr:paint()
+
+        cr:set_operator(cairo.Operator.CLEAR)
+        cr:set_line_width(2)
+        cr:stroke()
+    else
+        -- Fill the whole clip
+        cr:set_operator(cairo.Operator.SOURCE)
+        cr:set_source_rgba(1,1,1,1)
+        cr:paint()
+    end
+
+
+    self.shape_clip = img._native
+
+    -- Force the drawable to redraw the inner border
+    self:draw()
```

Comment:
Oh and this doesn't actually work. `draw()` will start a `delayed_call` to do a redraw. That call will then see "there is no need to re-layout widgets and the dirty area (which pixels in the wibox need to be redrawn?) is empty, so I don't have anything to do". Thus, this will not work.